### PR TITLE
multiple blindness modifications [CR]

### LINKED ISSUE #11558: Use fire extinguisher to blind enemies
Having lost a few of my early games to unhappy mooses, I was surprised to find that spraying enemies with a fire extinguisher doesn't seem to have any effect. Maybe allowing it to blind them would make the extinguishers more useful?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/9504793-use-fire-extinguisher-to-blind-enemies?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F146201&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by sparr**
A handful of things I thought of related to blindness. My first non-trivial patch here.

Closes #11558

### CODE REVIEW

**Comment by Coolthulhu** [comment]
I don't think blind players should be able to weld stuff - with or without goggles.
If this was the case, the survivors could just close their eyes when welding.

**Comment by sparr** [comment]
@Coolthulhu some welding tasks will already be prevented by the checks on crafting that prevent you from working on things you can't see. I think we should probably have more of those checks, and more skill penalties for blind/nearsighted/farsighted.

**Comment by sparr** [comment]
relatedly, maybe players should be allowed to try to weld without goggles, and get a long term blindness effect, or even a perception stat drop, if they do it...

**Comment by sparr** [comment]
@Coolthulhu also, to be clear, this patch doesn't let blind players do anything they couldn't already do

**Comment by Coolthulhu** [comment]
> relatedly, maybe players should be allowed to try to weld without goggles, and get a long term blindness effect, or even a perception stat drop, if they do it...

No permanent effects please. Those are a giant noob trap and don't add anything meaningful to the game.

> I think we should probably have more of those checks

In no case should blindness count as welding goggles. If a blind player can do something, a non-blind one should be able to do it too.

**Comment by sparr** [comment]
OK, well, let's leave the welding thing for another day. I won't be offended if that commit gets left behind.

**Comment by BevapDin** [comment]
> diff looks weird

https://github.com/blog/967-github-secrets should help with that.

**Comment by sparr** [comment]
Thanks @BevapDin 

Does anyone have feedback on the changes other than the welding goggles? I can do some commit wizardry to remove it, if the merge folks here are willing to take the other commits.

**Comment by Coolthulhu** [comment]
I'm not so sure about the field thing.
You can smell the gas/feel the heat of the fire/etc. before you stumble into it. Maybe change it to a move point penalty when the player tries to move into the field and then denies? Stumbling into fire is at the moment pretty much a death sentence for all but the most powerful survivors.

It would be cool to have an item that grants blindness when worn. Currently all player blindness is very short term, meaning that the part about blind characters not waking up when its bright will never happen.

Something more technical: you may want to consolidate those checks into `has_welding_goggles()` and `is_blind()` getters. That way you wouldn't have 5+ blocks of `if( x || y ||... )`.

**Comment by sparr** [comment]
@Coolthulhu a move point penalty when declining is a neat idea. I can definitely do that.

I realized that the blind wakeup thing might never trigger, but I included it for completeness. No idea if/when there will be permanent blindness (20 point trait?).

I already rolled back the welding goggle commit. If I brought it back, I might add a getter for it. Aside from that, I'm not sure if the difference between has_effect("blind") and is_blind() is big enough to warrant a getter.

**Comment by BevapDin** [comment]
> I'm not sure if the difference between has_effect("blind") and is_blind() is big enough to warrant a getter.

It will make implementing blindness as a trait or similar easy. It will also allow items that give that effect. Look at the implementation of `player::is_deaf`, code that needs to check for the character being deaf does not care whether it is from a worn item or from an effect.

**Comment by sparr** [comment]
Ahh, that makes sense. I don't know that I'm ready to tackle that sort of design/architecture change yet. I was trying to get my feet wet with simple features, first.

**Comment by Coolthulhu** [comment]
It's trivial actually: just add an is_deaf() function in player class and implement it.

It kinda goes beyond the scope of this PR, but if you later want to expand on the whole blindness thing, it would be a good idea to add it as soon as there is more than one possible kind of blindness (like, permanent vs temporary).

**Comment by sparr** [comment]
@Coolthulhu just adding is_blind to Player isn't everything, though. the blind effect is checked in Character and Monster, as well. I'm thinking is_blind might belong all the way up in Creature, and is_deaf maybe too. That's the sort of architecture decision I don't feel comfortable making on this project yet.

**Comment by KA101** [comment]
>It's trivial actually: just add an is_deaf() function in player class and implement it.

It's trivial for Coolthulhu, sure; not necessarily for others.  ;-)

**Comment by Coolthulhu** [comment]
It's just a virtual function. I'm pretty sure sparr is a bit too humble here and could handle it just fine.
It's not that it's an important change (it kinda goes beyond the scope of this PR to be honest), just that fear of breaking things or doing them wrong should not stop people from making small changes like this. Especially since small PRs are easy to review.

This community isn't hostile towards people who make mistakes. Otherwise I'd be kicked out long ago.

Broken eggs, omelets etc.
BTW We have no omelet recipe.

**Comment by sparr** [comment]
@Coolthulhu my concern is that if I put is_blind in Player then there's still a lot of code duplication in Character and Monster. But if I put is_blind in Creature then I can't check monster flags. Yes, there are a bunch of simple approaches here, but none of them are the right approach, and someone who knows how all the mob classes interrelate has to make some of those decisions. Give me a few weeks; if I'm still contributing then then I'll try to tackle is_blind().

**Comment by Coolthulhu** [comment]
@sparr Sure, no rush. Not saying that it has to be done now or by you (or even at all), just that fear of breaking things shouldn't be an issue here.

**Comment by sparr** [comment]
gonna mark this non-WIP and see if it can get merged, and I'll work from master after it does.

File: src/game.cpp [hunk 1]
```diff
@@ -11710,43 +11710,45 @@ bool game::plmove(int dx, int dy)
...
-                              (u.get_env_resist(bp_eyes) >= 15) ) &&
-                              !u.has_trait("M_IMMUNE"));
-                break;
-            default:
-                dangerous = cur.is_dangerous();
-                break;
-            }
-            if ((dangerous) && !query_yn(_("Really step into that %s?"), cur.name().c_str())) {
-                return false;
+        if (!u.has_effect("blind")) {
```
Comment (KA101) [inline src/game.cpp]: Lots of these would likely be detectable by smell, possibly (definitely in the case of smoke and tear gas) before you got into the field.  I'm not sold on the blanket exclusion.

Comment (sparr) [inline src/game.cpp]: @KA101 is there some way to impose a partial penalty? maybe check the field strength in the target square and impose a 1/2 strength effect if the player declines to continue.

when the player is standing adjacent to a smoke or gas field, sure, they can smell the smoke or gas... but they don't know when walking whether they are headed in or out of the field until it's too late.

**Comment by KA101** [comment]
Probably, but at the end of the day Blinded is harsh enough IMO and my nose, stuffed up as it may get, still lets me know about smoke from a few city blocks (longer than a house-length) away.  More interested if blind & wearing partial-protection, which would block smell.

**Comment by sparr** [comment]
@KA101 My point is that knowing about smoke from a block away isn't relevant. Your nose tells you there is smoke, not where it is. When you are standing in an almost-smoky-enough-to-be-a-problem tile, you're just one meter away from a barely-smoky-enough-to-be-a-problem tile. Yes, obviously you can smell the smoke from there, or from 10 tiles away (if we care to implement sense of smell at some point in the fuure, which I think would be awesome). But you can't tell which way leads to more smoke and which way leads to less smoke.

Anyway, the smell thing is a fun idea, as is mixing smell and sight, and I can revisit that later. I've just committed a partial revert, so now this PR only hides traps from blind players, not smoke/gas/etc.

**Comment by KA101** [comment]
>Can't determine direction to smoke based on smell

My nose very much begs to differ, sorry.  It's quite possible that I'm more sensitive than normal, and I'd anticipate that blind folks would have or develop similar sensitivity to compensate, given time.  So worth considering for longer-term handling.

**Comment by KA101** [comment]
OK, and if need be I can whip up a blindfold.  Honestly surprised nobody else has, given that we have earplugs that do the same thing for deafness.

**Comment by DavidKeaton** [comment]
Even as a smoker (sadly) I can tell when smoke is "thicker" but not where to go to find that scent.

**Pull request merged**